src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1577,6 +1595,66 @@ public override BoundNode VisitCall(BoundCall node)
             return null;
         }
 
+        private bool MethodReturnsTrueWhenNotNull(MethodSymbol method)
+        {
+            if (method.ReturnType.SpecialType != SpecialType.System_Boolean)
+            {
+                return false;
+            }
+
+            return method.Equals(compilation.GetWellKnownTypeMember(WellKnownMember.System_String_IsNullOrEmpty));
+        }
+
+        private void VisitTrueWhenNotNull(BoundExpression operand, TypeSymbol boolType)
+        {
+            if (!operand.Type.IsReferenceType)
+            {
+                return;
+            }
+
+            // operand == null || someUnknownBoolean
+            BoundExpression equivalent = LogicalOr(
+                new BoundBinaryOperator(operand.Syntax, BinaryOperatorKind.Equal, operand,
+                    new BoundLiteral(operand.Syntax, ConstantValue.Null, operand.Type),
+                    constantValueOpt: null, methodOpt: null, LookupResultKind.Viable, boolType),
+                new BoundLiteral(operand.Syntax, constantValueOpt: null, boolType));
+
+            // We only need to update the null-state, so we'll do the flow analysis without diagnostics
+            bool savedDisableDiagnostics = _disableDiagnostics;
+            _disableDiagnostics = true;
+
+            Visit(equivalent);
+
+            _disableDiagnostics = savedDisableDiagnostics;
+        }
+
+        internal static BoundBinaryOperator LogicalOr(BoundExpression left, BoundExpression right)
+        {
+            return new BoundBinaryOperator(left.Syntax, BinaryOperatorKind.LogicalBoolOr, left, right, constantValueOpt: null, methodOpt: null, LookupResultKind.Viable, left.Type);
+        }
+
+        private bool MethodEnsuresTrueWhenExits(MethodSymbol method)
+        {
+            return method.Equals(compilation.GetWellKnownTypeMember(WellKnownMember.System_Diagnostics_Debug_Assert1)) ||
+                method.Equals(compilation.GetWellKnownTypeMember(WellKnownMember.System_Diagnostics_Debug_Assert2)) ||
+                method.Equals(compilation.GetWellKnownTypeMember(WellKnownMember.System_Diagnostics_Debug_Assert3)) ||
+                method.Equals(compilation.GetWellKnownTypeMember(WellKnownMember.System_Diagnostics_Debug_Assert4));
```

Comment:
Why use `WellKnownMember` rather than the extra annotations table? (Presumably, we won't need these well-known members once we support annotations.)

Reply:
What we discussed in LDM was that some members would be well-known. That said, maybe that was only for built-in operators?
I like the extra annotations approach. I'll try that.